src/main/scala/io/picnicml/doddlemodel/preprocessing/QuantileDiscretizer.scala

```diff
@@ -0,0 +1,106 @@
+
+case class QuantileDiscretizer(
+  private val bucketCounts: DenseVector[Double],
+  private val featureIndex: FeatureIndex,
+  private val quantiles: Option[Seq[Seq[(Double, Double)]]] = None
+) {
+  private val numNumeric = featureIndex.numerical.columnIndices.length
+  require(numNumeric == 0 || numNumeric == bucketCounts.length, "A quantile should be given for every numerical column")
+}
+
+/** An immutable preprocessor that discretizes numerical features into buckets based on quantiles.
+  * Numerical feature values are converted to categorical features based on their distribution
+  * */
+object QuantileDiscretizer {
+
+  /** Create a quantile discretizer which splits data into discrete evenly sized buckets.
+    *
+    * @param bucketCount The number of quantiles desired
+    * @param featureIndex feature index associated with features - this is needed so that only numerical features are
+    *                     transformed by this preprocessor; could be a subset of columns to be transformed
+    *
+    * @example Discretize a matrix into quartiles with two features: one numerical and one categorical.
+    *   {{{
+    *     import io.picnicml.doddlemodel.preprocessing.QuantileDiscretizer.ev
+    *     import io.picnicml.doddlemodel.syntax.TransformerSyntax._
+    *
+    *     val featureIndex = FeatureIndex(List(NumericalFeature, CategoricalFeature))
+    *     val x = DenseMatrix(
+    *       List(-1.0, 0.0),
+    *       List(0.0, 1.0),
+    *       List(2.0, 0.0),
+    *       List(5.0, 0.0)
+    *     )
+    *     // equivalently, DenseVector(4) could be used
+    *     val bucketCounts = 4
+    *     val discretizer = QuantileDiscretizer(bucketCounts, featureIndex).fit(x)
+    *     val xTransformed = discretizer.transform(x)
+    *   }}}
+    */
+  def apply(bucketCount: Int, featureIndex: FeatureIndex): QuantileDiscretizer = {
+    val numNumeric = featureIndex.numerical.columnIndices.length
+    val bucketCountExtended: DenseVector[Double] = DenseVector.fill(numNumeric) { bucketCount.toDouble }
+    QuantileDiscretizer(bucketCountExtended, featureIndex)
+  }
+
+  implicit lazy val ev: Transformer[QuantileDiscretizer] = new Transformer[QuantileDiscretizer] {
+
+    @inline override def isFitted(model: QuantileDiscretizer): Boolean = model.quantiles.isDefined
+
+    override def fit(model: QuantileDiscretizer, x: Features): QuantileDiscretizer = {
+      val discreteRangeArrays = model.featureIndex.numerical.columnIndices.zipWithIndex.map {
+        case (colIndex, bucketCountsIndex) =>
+          val colArray = x(::, colIndex).toScalaVector.sorted
+          computeQuantiles(colArray, model.bucketCounts(bucketCountsIndex).toInt)
+      }
+      model.copy(quantiles = discreteRangeArrays.some)
+    }
+
+    override protected def transformSafe(model: QuantileDiscretizer, x: Features): Features = {
+      val xCopy = x.copy
+      model.featureIndex.numerical.columnIndices.zipWithIndex.foreach {
```

Comment:
I like this, it's elegant. I personally would format it as (just a subjective preference):
```scala
model.featureIndex.numerical.columnIndices.zipWithIndex.foreach { case (colIndex, bucketsIndex) =>
  val buckets = model.quantiles.getOrBreak(bucketsIndex)
  (0 until xCopy.rows).foreach { rowIndex =>
    xCopy(rowIndex, colIndex) = buckets.indexWhere { case (lowerBound, upperBound) =>
      lowerBound <= xCopy(rowIndex, colIndex) && xCopy(rowIndex, colIndex) <= upperBound
    }.toDouble
  }
}
```